x-pack/plugin/profiler/src/main/java/org/elasticsearch/xpack/profiler/AbstractProfilingPersistenceManager.java

```diff
@@ -0,0 +1,226 @@
+import org.elasticsearch.cluster.metadata.IndexMetadata;
+import org.elasticsearch.cluster.metadata.MappingMetadata;
+import org.elasticsearch.cluster.routing.IndexRoutingTable;
+import org.elasticsearch.cluster.service.ClusterService;
+import org.elasticsearch.gateway.GatewayService;
+import org.elasticsearch.index.Index;
+
+import java.io.Closeable;
+import java.util.Map;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+public abstract class AbstractProfilingPersistenceManager<T extends AbstractProfilingPersistenceManager.ProfilingIndexAbstraction>
+    implements
+        ClusterStateListener,
+        Closeable {
+    protected final Logger logger = LogManager.getLogger(getClass());
+
+    private final AtomicBoolean inProgress = new AtomicBoolean(false);
+
+    private final ClusterService clusterService;
+    private volatile boolean templatesEnabled;
+
+    public AbstractProfilingPersistenceManager(ClusterService clusterService) {
+        this.clusterService = clusterService;
+    }
+
+    public void initialize() {
+        clusterService.addListener(this);
+    }
+
+    @Override
+    public void close() {
+        clusterService.removeListener(this);
+    }
+
+    public void setTemplatesEnabled(boolean templatesEnabled) {
+        this.templatesEnabled = templatesEnabled;
+    }
+
+    @Override
+    public final void clusterChanged(ClusterChangedEvent event) {
+        if (templatesEnabled == false) {
+            return;
+        }
+        // wait for the cluster state to be recovered
+        if (event.state().blocks().hasGlobalBlock(GatewayService.STATE_NOT_RECOVERED_BLOCK)) {
+            return;
+        }
+
+        // If this node is not a master node, exit.
+        if (event.state().nodes().isLocalNodeElectedMaster() == false) {
+            return;
+        }
+
+        if (event.state().nodes().getMaxNodeVersion().after(event.state().nodes().getSmallestNonClientNodeVersion())) {
+            logger.debug("Skipping up-to-date check as cluster has mixed versions");
+            return;
+        }
+
+        if (isAllResourcesCreated(event) == false) {
+            logger.trace("Skipping index creation; not all required resources are present yet");
+            return;
+        }
+
+        if (inProgress.compareAndSet(false, true)) {
+            // Only release the lock once all upgrade attempts have succeeded or failed.
+            try (var refs = new RefCountingRunnable(() -> inProgress.set(false))) {
+                ClusterState clusterState = event.state();
+                for (T index : getManagedIndices()) {
+                    Status status = getStatus(clusterState, index);
+                    if (status.actionable) {
+                        onStatus(clusterState, status, index, ActionListener.releasing(refs.acquire()));
+                    }
+                }
+            }
+        } else {
+            logger.trace("Skipping index creation as changes are already in progress");
+        }
```

Comment:
```suggestion
        if (inProgress.compareAndSet(false, true) == false) {
            logger.trace("Skipping index creation as changes are already in progress");
            return
        }

        // Only release the lock once all upgrade attempts have succeeded or failed.
        try (var refs = new RefCountingRunnable(() -> inProgress.set(false))) {
            ClusterState clusterState = event.state();
            for (T index : getManagedIndices()) {
                Status status = getStatus(clusterState, index);
                if (status.actionable) {
                    onStatus(clusterState, status, index, ActionListener.releasing(refs.acquire()));
                }
            }
        }
```

Reply:
Good point. Done in https://github.com/elastic/elasticsearch/pull/97380/commits/5d88466207834582b22c8227aacc8dbb88413ac7.